File: src/frontend/src/lib/sensorTrend.ts
import type { SensorReading } from "../api/types";

export type TrendDirection = "up" | "down";

/**
 * Compares the average of the first third of readings against the last
 * third to determine trend direction, smoothing out day-to-day noise.
 * Returns whether the trend is moving toward the "bad" direction for
 * that metric (e.g. rising temperature, falling oil quality).
 */
export function isDegrading(readings: SensorReading[], badDirection: TrendDirection): boolean {
  if (readings.length < 2) return false;

  const chunk = Math.max(1, Math.floor(readings.length / 3));
  const early = readings.slice(0, chunk);
  const late = readings.slice(-chunk);

  const avg = (arr: SensorReading[]) => arr.reduce((sum, r) => sum + r.value, 0) / arr.length;
  const delta = avg(late) - avg(early);

  return badDirection === "up" ? delta > 0 : delta < 0;
}

File: src/mobile/lib/util/sensor_trend.dart
import '../api/types.dart';

enum TrendDirection { up, down }

/// Port of the web frontend's `src/lib/sensorTrend.ts`.
///
/// Compares the average of the first third of readings against the last third
/// to determine trend direction, smoothing out day-to-day noise. Returns
/// whether the trend is moving toward the "bad" direction for that metric
/// (e.g. rising temperature, falling oil quality).
bool isDegrading(List<SensorReading> readings, TrendDirection badDirection) {
  if (readings.length < 2) return false;

  final chunk = (readings.length ~/ 3).clamp(1, readings.length);
  final early = readings.take(chunk);
  final late = readings.skip(readings.length - chunk);

  double avg(Iterable<SensorReading> items) =>
      items.fold<double>(0, (sum, r) => sum + r.value) / items.length;

  final delta = avg(late) - avg(early);

  return badDirection == TrendDirection.up ? delta > 0 : delta < 0;
}
